android/src/main/java/com/wix/reactnativenotifications/core/JsIOHelper.java

```diff
@@ -1,38 +1,89 @@
 package com.wix.reactnativenotifications.core;
 
 import android.content.Context;
+import android.content.Intent;
 import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
 
 import com.facebook.react.ReactApplication;
+import com.facebook.react.ReactInstanceManager;
+import com.facebook.react.ReactNativeHost;
 import com.facebook.react.bridge.Arguments;
 import com.facebook.react.bridge.ReactContext;
-import com.facebook.react.bridge.WritableMap;
 import com.facebook.react.modules.core.DeviceEventManagerModule;
 
+import java.util.LinkedList;
+
 public class JsIOHelper {
 
+    private static final LinkedList<Event> sBackgroundQueue = new LinkedList<Event>();
+    private static boolean sJsIsReady = false;
+
     private final Context mContext;
+    private final Handler mHandler = new Handler(Looper.getMainLooper());
 
     public JsIOHelper(Context context) {
         mContext = context;
     }
 
-    public boolean sendEventToJS(String eventName, Bundle bundle) {
-        return sendEventToJS(eventName, Arguments.fromBundle(bundle));
+    public void sendEventToJS(String name, Bundle bundle) {
+        final Event event = new Event(name, Arguments.fromBundle(bundle));
+        postEvent(event);
     }
 
-    public boolean sendEventToJS(String eventName, String string) {
-        return sendEventToJS(eventName, (Object) string);
+    public void sendEventToJS(String name, String string) {
+        final Event event = new Event(name, string);
+        postEvent(event);
     }
 
-    boolean sendEventToJS(String eventName, Object object) {
-        final ReactContext reactContext = ((ReactApplication) mContext.getApplicationContext()).getReactNativeHost().getReactInstanceManager().getCurrentReactContext();
+    public void consumeBackgroundQueue() {
+        synchronized (JsIOHelper.class) {
+            sJsIsReady = true;
 
-        if (reactContext != null && reactContext.hasActiveCatalystInstance()) {
-            reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(eventName, object);
-            return true;
+            while (!sBackgroundQueue.isEmpty()) {
+                emitEvent(sBackgroundQueue.pop());
+            }
         }
+    }
+
+    private void postEvent(final Event event) {
+        if (Looper.getMainLooper() == Looper.myLooper()) {
+            emitEvent(event);
+        } else {
+            mHandler.post(new Runnable() {
```

Comment:
If anybody is wondering why this is necessary, it is because our events are coming from `IntentService` (not regular `Service`) sub-classes. `IntentService` run on a dedicated background thread rather than the main thread. This becomes a problem when the app is launched in response to push notification being received. In particular, no `ReactInstanceManager`exists yet, and attempting to retrieve/create it leads to a crash inside React Native. React Native expects `ReactInstanceManager` to only be _created_ on the main thread (although it can be retrieved and utilised from any thread), internally attempting to create an Android `Handler` on an "unprepared" `Looper` leads to a runtime exception.